File: backend/mcp-code-intelligence-python/src/mcp_code_intel/analyzers/security/injection/pattern_matcher.py
"""
KSA-165: Pattern Matcher — Base class for injection pattern matching.
"""

from __future__ import annotations
from abc import ABC, abstractmethod
from typing import Optional
from ..types import TaintPath, InjectionPattern, Finding


class MatchContext:
    __slots__ = ("file_path", "function_name", "language")

    def __init__(self, file_path: str, function_name: str, language: str):
        self.file_path = file_path
        self.function_name = function_name
        self.language = language


class PatternMatcher(ABC):
    @property
    @abstractmethod
    def category(self) -> str: ...

    @property
    @abstractmethod
    def patterns(self) -> list[InjectionPattern]: ...

    def match(self, taint_path: TaintPath, context: MatchContext) -> Optional[Finding]:
        """Check if a taint path matches any pattern in this category."""
        for pattern in self.patterns:
            if (self._matches_sink(taint_path.sink.function, pattern) and
                self._has_dangerous_op(taint_path, pattern.dangerous_ops) and
                not self._has_safe_pattern(taint_path, pattern.safe_patterns)):
                return self._create_finding(taint_path, pattern, context)
        return None

    def _matches_sink(self, sink_function: str, pattern: InjectionPattern) -> bool:
        return any(sp in sink_function for sp in pattern.sink_patterns)

    def _has_dangerous_op(self, path: TaintPath, dangerous_ops: list[str]) -> bool:
        if not dangerous_ops:
            return True
        return any(step.action in dangerous_ops for step in path.chain)

    def _has_safe_pattern(self, path: TaintPath, safe_patterns: list[str]) -> bool:
        if not safe_patterns:
            return False
        sink_expr = path.sink.expression
        for safe in safe_patterns:
            if safe in sink_expr:
                return True
        return any(step.action == "sanitize" for step in path.chain)

    def _create_finding(self, path: TaintPath, pattern: InjectionPattern, context: MatchContext) -> Finding:
        confidence = self._compute_confidence(path, pattern)
        return Finding(
            id=f"{pattern.category.upper()}-{pattern.id}-{context.file_path}:{path.sink.line}",
            rule_id=f"INJ-{pattern.category.upper()}-{str(pattern.id).zfill(3)}",
            category=pattern.category,
            pattern=pattern,
            taint_path=path,
            severity=pattern.severity,
            confidence=confidence,
            cwe=pattern.cwe,
            message=f"{pattern.name}: Tainted data from {path.source.type} flows to {path.sink.function} without sanitization",
            remediation=pattern.description,
            location={"file": context.file_path, "start_line": path.source.line, "end_line": path.sink.line},
            suppressed=False,
        )

    def _compute_confidence(self, path: TaintPath, pattern: InjectionPattern) -> str:
        if path.length <= 3:
            return "High"
        if path.length <= 6:
            return "Medium"
        return "Low"

File: backend/mcp-code-intelligence-kotlin/src/main/kotlin/com/codeintel/analyzers/security/injection/PatternMatcher.kt
/**
 * KSA-165: Pattern Matcher — Base class for injection pattern matching.
 */
package com.codeintel.analyzers.security.injection

import com.codeintel.analyzers.security.*

data class MatchContext(val filePath: String, val functionName: String, val language: String)

abstract class PatternMatcher {
    abstract val category: String
    abstract val patterns: List<InjectionPattern>

    fun match(taintPath: TaintPath, context: MatchContext): Finding? {
        for (pattern in patterns) {
            if (matchesSink(taintPath.sink.function, pattern) &&
                hasDangerousOp(taintPath, pattern.dangerousOps) &&
                !hasSafePattern(taintPath, pattern.safePatterns)) {
                return createFinding(taintPath, pattern, context)
            }
        }
        return null
    }

    private fun matchesSink(sinkFunction: String, pattern: InjectionPattern): Boolean =
        pattern.sinkPatterns.any { it in sinkFunction }

    private fun hasDangerousOp(path: TaintPath, dangerousOps: List<String>): Boolean {
        if (dangerousOps.isEmpty()) return true
        return path.chain.any { it.action.name.lowercase() in dangerousOps }
    }

    private fun hasSafePattern(path: TaintPath, safePatterns: List<String>): Boolean {
        if (safePatterns.isEmpty()) return false
        val sinkExpr = path.sink.expression
        if (safePatterns.any { it in sinkExpr }) return true
        return path.chain.any { it.action == TaintStepAction.SANITIZE }
    }

    private fun createFinding(path: TaintPath, pattern: InjectionPattern, context: MatchContext): Finding {
        val confidence = if (path.length <= 3) Confidence.HIGH else if (path.length <= 6) Confidence.MEDIUM else Confidence.LOW
        return Finding(
            id = "${pattern.category.uppercase()}-${pattern.id}-${context.filePath}:${path.sink.line}",
            ruleId = "INJ-${pattern.category.uppercase()}-${pattern.id.toString().padStart(3, '0')}",
            category = pattern.category,
            pattern = pattern,
            taintPath = path,
            severity = pattern.severity,
            confidence = confidence,
            cwe = pattern.cwe,
            message = "${pattern.name}: Tainted data from ${path.source.type} flows to ${path.sink.function} without sanitization",
            remediation = pattern.description,
            location = mapOf("file" to context.filePath, "startLine" to path.source.line, "endLine" to path.sink.line)
        )
    }
}
